mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -1307,6 +1308,27 @@ def _get_sqlalchemy_filter_clauses(parsed, session, dialect):
                 entity = SqlParam
             elif SearchUtils.is_tag(key_type, comparator):
                 entity = SqlTag
+            elif SearchUtils.is_condition(key_type, comparator):
+                f = SearchUtils.get_sql_comparison_func(comparator, dialect)
+
+                _sub_attribute_filters, _sub_non_attribute_filters = _get_sqlalchemy_filter_clauses(
+                    value, session, dialect
+                )
+
+                if comparator == "AND":
+                    attribute_filters.extend(_sub_attribute_filters)
+                    non_attribute_filters.extend(_sub_non_attribute_filters)
+                elif comparator == "OR":
+                    if _sub_attribute_filters:
+                        attribute_filters.append(
+                            f("", [q.element for q in _sub_attribute_filters]).subquery()
+                        )
+
+                    if _sub_non_attribute_filters:
+                        non_attribute_filters.append(
+                            f("", [q.element for q in _sub_non_attribute_filters]).subquery()
+                        )
+                continue
```

Comment:
The logic makes the generated sql very complex. Can we do some refactor first and then we can simplify the logic here:

1)
`_get_sqlalchemy_filter_clauses` it only parses user-input filter string, and returns sqlalchemy filter .

2)
we traverse the sqlalchemy filter generated by step-(1), and generate the list of database tables that the filter uses.

3) generate query by:
* first we join all tables generated by step (2)
* apply the filter generated by step (1) on top of the joined table query.

CC @dbczumar WDYT ?

Similarly we can refactor and simplify other sql related code for experiment 
 / model model versions searching, and we can make them share some utility code. Otherwise current code we have many complicated and similar logic in many places and make the code hard to maintain.

Reply:
Makes sense to me!

Reply:
yeah same, We can do it as a separate module for now to avoid using the production code right away. This will prob take multiple PRS anyways so keeping separate will allow others to take over or jump in at any point. The diff will be easier to read too.

Reply:
The other crazy thought I had was unifiying the file search and sqlalchemy search by using filestore and SQLite temp tables to load the page results into and then just using single sqlalchemy parsing. But that’s more work than we want to.